README.md
# fecR

fecR is a package for calculating fishing effort. It is based on work carried out at the 2nd DG MARE Ad-Hoc Working Group on Transversal Variables, Nicosia, Cyprus 22-26 February 2016.

Two measure of fishing effort are calculated: *Days at Sea* and *Fishing Days*.

There are two vignettes in the package. One to show you how to prepare your data and the either to show you how to calculate fishing effort.

This package is offered in good faith and with no guarantees.

# References

Castro Ribeiro, C., Holmes, S., Scott, F., Berkenhagen, J., Demaneche, S., Prista, N., Reis, D., Reilly,
T., Andriukaitiene, J., Aquilina, M., Avdič Mravlje, E., Calvo Santos, A., Charilaou, C., Dalskov, J., Davidiuk, I.,
Diamant, A., Egekvist, J., Elliot, M., Ioannou, M., Jakovleva, I. Kuzebski, E., Ozernaja, O., Pinnelo, D., Thasitis,
I., Verlé, K., Vitarnen, J., Wójcik, I..Report of the 2nd Workshop on Transversal Variables. Nicosia, Cyprus.
22-26 February 2016. A DCF ad-hoc workshop EUR; doi



NEWS.md
# fecR 0.0.2

## New features

* Fixed bug about duplicate trips even though different vessels

# fecR 0.0.1

## New features

* All of them. It is the first release!



tests/testthat.R
library(testthat)
library(fecR)

test_check("fecR")


tests/testthat/make_okdata.R
# Author: Finlay Scott (EC JRC) <iago.mosqueira@jrc.ec.europa.eu>, Nuno Prista (SLU) and Thomas Reilly (Marine Scotland)
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

okdata <- data.frame(
    eunr_id = "my_boat", loa = 2000, gt = 70, kw = 400,
    trip_id = c("trip1","trip1","trip2","trip2"),
    depdate = rep(c("20140718", "20141023"), each=2),
    deptime = rep(c("0615", "0730"), each=2),
    retdate = rep(c("20140719", "20141024"), each=2),
    rettime = rep(c("1830", "1615"), each=2),
    fishdate = c("20140718", "20140719", "20141023", "20141024"),
    gear = c("OTB","OTB","GN","GN"), gear_mesh_size = 80, fishing_area = "27.4.B",
    economic_zone = "EU", rectangle = "39F0",
    stringsAsFactors = FALSE
)


tests/testthat/test-check_format.R
# Author: Finlay Scott (EC JRC) <iago.mosqueira@jrc.ec.europa.eu>, Nuno Prista (SLU) and Thomas Reilly (Marine Scotland)
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

# Check the check_format function
library(fecR)
context("check_format")
source("make_okdata.R")

test_that("ok data", {
    # Should pass without warning
    expect_warning(out <- check_format(okdata), regexp=NA)
    expect_equal(out, okdata)
})

test_that("extra column", {
    extra_col <- cbind(okdata, new_col = runif(nrow(okdata)))
    expect_warning(out <- check_format(extra_col))
    expect_warning(corout <- check_format(extra_col, correct=TRUE), regexp=NA)
    # Try corrected data
    expect_warning(out <- check_format(corout), regexp=NA)
})

test_that("wrong column names", {
    wrong_col <- okdata
    colnames(wrong_col)[3] <- "something"
    expect_warning(out <- check_format(wrong_col))
})

test_that("eurn_id column", {
    wrong_eunr_id <- okdata
    # Force them to be numeric instead of character
    wrong_eunr_id[,"eunr_id"] <- 1234
    expect_warning(out <- check_format(wrong_eunr_id))
    expect_warning(corout <- check_format(wrong_eunr_id, correct=TRUE), regexp=NA)
    # Check the corrected output
    expect_warning(out <- check_format(corout), regexp=NA)
    wrong_eunr_id <- okdata
    # Set entry to be missing
    wrong_eunr_id[1,"eunr_id"] <- as.character(NA)
    expect_warning(out <- check_format(wrong_eunr_id))
    # No correction possible
    expect_warning(out <- check_format(wrong_eunr_id, correct=TRUE))
})

test_that("loa, gt and kw", {
    for (col in c("loa","gt","kw")){
        wrong_col <- okdata
        rws <- sample(1:nrow(wrong_col), runif(1,1,nrow(wrong_col)))
        # Make character with numeric
        wrong_col[rws,col] <- "90m"
        expect_warning(out <- check_format(wrong_col))
        expect_warning(corout <- check_format(wrong_col, correct=TRUE), regexp=NA)
        expect_warning(out <- check_format(corout), regexp=NA)
        # Make character without numeric - no correction
        wrong_col[rws,col] <- "ninety"
        expect_warning(out <- check_format(wrong_col))
        expect_warning(corout <- check_format(wrong_col, correct=TRUE))
        expect_warning(out <- check_format(corout))
        # If missing
        wrong_col <- okdata
        wrong_col[rws,col] <- NA
        expect_warning(out <- check_format(wrong_col))
    }
})

test_that("date columns", {
    for(date_col in c("depdate","retdate", "fishdate")){
        wrong_date <- okdata
        # Needs to be character string, not numeric even if format is OK
        wrong_date[,date_col] <- as.numeric(wrong_date[,date_col])
        expect_warning(out <- check_format(wrong_date))
        # We can correct
        expect_warning(corout <- check_format(wrong_date, correct=TRUE), regexp=NA)
        expect_warning(out <- check_format(corout), regexp=NA)
        # If the format is wrong then we cannot automatically correct and check complains.
        wrong_date <- okdata
        rws <- sample(1:nrow(wrong_date), runif(1,1,nrow(wrong_date)))
        # Wrong format - year is too short
        wrong_date[rws, date_col] <- "141024"
        expect_warning(out <- check_format(wrong_date))
        expect_warning(out <- check_format(wrong_date, correct=TRUE))
        # Wrong format again - month must be a numeric character
        wrong_date[rws, date_col] <- "October14"
        expect_warning(out <- check_format(wrong_date))
        expect_warning(out <- check_format(wrong_date, correct=TRUE))
        # Missing data is not allowed and check complains.  It is not possible to correct for missing data.
        wrong_date <- okdata
        wrong_date[rws, date_col] <- as.character(NA)
        expect_warning(out <- check_format(wrong_date))
        expect_warning(out <- check_format(wrong_date, correct=TRUE))
    }
})

test_that("time column", {
    for (time_col in c("deptime", "rettime")){
        wrong_time <- okdata
        # Needs to be character string, not numeric even if format is OK
        wrong_time[, time_col] <- 1000
        expect_warning(out <- check_format(wrong_time))
        # We can correct by forcing to character
        expect_warning(corout <- check_format(wrong_time, correct=TRUE), regexp=NA)
        expect_warning(out <- check_format(corout), regexp=NA)
        # If only 2 characters are provided and automatic correction is TRUE,
        # the characters are assumed to be hours (*HH*) and minutes of "00" 
        # are appended to the string, e.g. "16" becomes "1600".
        wrong_time <- okdata
        wrong_time[wrong_time$trip_id=="trip1", time_col] <- "16"
        expect_warning(out <- check_format(wrong_time))
        expect_warning(outcor <- check_format(wrong_time, correct=TRUE), regexp=NA)
        expect_warning(out <- check_format(outcor), regexp=NA)
        # This automatic correction only works if the 2 characters are numeric
        wrong_time <- okdata
        wrong_time[wrong_time$trip_id=="trip1", time_col] <- "TT"
        expect_warning(out <- check_format(wrong_time))
        expect_warning(out <- check_format(wrong_time, correct=TRUE))
        # Two or four characters are needed. For example, "0130" is OK whereas "130" is not
        wrong_time <- okdata
        wrong_time[wrong_time$trip_id=="trip1", time_col] <- "130"
        expect_warning(out <- check_format(wrong_time))
        expect_warning(out <- check_format(wrong_time, correct=TRUE))
        # A : separator is accecptable (*HH:MM*) but is removed from the returned data.
        wrong_time <- okdata
        wrong_time[wrong_time$trip_id=="trip1", time_col] <- "16:15"
        expect_warning(out <- check_format(wrong_time), regexp=NA)
        expect_equal(out[out$trip_id=="trip1", time_col], rep("1615", sum(out$trip_id=="trip1")))
        # *NA*s and missing values are not acceptable and cannot be automatically corrected.
        wrong_time <- okdata
        wrong_time[wrong_time$trip_id=="trip1", time_col] <- as.character(NA)
        expect_warning(out <- check_format(wrong_time))
        expect_warning(out <- check_format(wrong_time, correct=TRUE))
    }
})

test_that("gear codes", {
    #Whitespace is not allowed. It can be automatically removed if wanted.
    wrong_gear <- okdata
    wrong_gear[1,"gear"] <- " OTB"
    expect_warning(out <- check_format(wrong_gear))
    # Correct removes whitespace
    expect_warning(corout <- check_format(wrong_gear, correct=TRUE), regexp=NA)
    expect_warning(out <- check_format(corout), regexp=NA)
    # If the gear code is not found in the MDR list then no automatic correction is possible.
    # Here the gear is unknown because it is lower case. All gear codes must be upper case.
    wrong_gear[1,"gear"] <- "otb"
    expect_warning(out <- check_format(wrong_gear))
})

test_that("gear mesh size", {
    # Entries that are not integer will make check complain. There is no option to autocorrect this.
    wrong_ms <- okdata
    # Text in the entry - must be integer - no correction possible
    row <- runif(1,1,nrow(wrong_ms))
    wrong_ms[row,"gear_mesh_size"] <- "80mm"
    expect_warning(out <- check_format(wrong_ms))
    expect_warning(out <- check_format(wrong_ms, correct=TRUE))
    # Cannot be an integer
    wrong_ms[row,"gear_mesh_size"] <- 80.8
    expect_warning(out <- check_format(wrong_ms))
    expect_warning(out <- check_format(wrong_ms, correct=TRUE))
    # If an entry is missing (e.g. it is *NA*) then check will complain.
    # It is possible to automatically correct this in which case the missing entry has a mesh size of 0.
    # The returned data will pass check but may not be what you want.
    wrong_ms[row,"gear_mesh_size"] <- NA
    expect_warning(out <- check_format(wrong_ms))
    expect_warning(corout <- check_format(wrong_ms, correct=TRUE) , regexp=NA)
    expect_warning(out <- check_format(corout), regexp=NA)
    expect_equal(corout[row,"gear_mesh_size"], 0L)
})

test_that("fishing area", {
    # Can correct for some errors
    wrong_fish_area <- okdata
    # Point at end - bad syntax
    row <- runif(1,1,nrow(wrong_fish_area))
    wrong_fish_area[row,"fishing_area"] <- "27.4.A."
    expect_warning(out <- check_format(wrong_fish_area))
    expect_warning(corout <- check_format(wrong_fish_area, correct=TRUE), regexp=NA)
    expect_warning(out <- check_format(corout), regexp=NA)
    # Lowercase
    wrong_fish_area[row,"fishing_area"] <- "27.4.a"
    expect_warning(out <- check_format(wrong_fish_area))
    expect_warning(corout <- check_format(wrong_fish_area, correct=TRUE), regexp=NA)
    expect_warning(out <- check_format(corout), regexp=NA)
    # White space
    wrong_fish_area[row,"fishing_area"] <- "27.4.A "
    expect_warning(out <- check_format(wrong_fish_area))
    expect_warning(corout <- check_format(wrong_fish_area, correct=TRUE), regexp=NA)
    expect_warning(out <- check_format(corout), regexp=NA)
    # Missing values are not allowed and cannot be automatically corrected.
    wrong_fish_area[row,"fishing_area"] <- as.character(NA)
    expect_warning(out <- check_format(wrong_fish_area))
    expect_warning(out <- check_format(wrong_fish_area, correct=TRUE))
})

test_that("economic zone", {
    # Wrong code - cannot be corrected
    wrong_econ <- okdata
    wrong_econ[3,"economic_zone"] <- "USA"
    expect_warning(out <- check_format(wrong_econ))
    expect_warning(out <- check_format(wrong_econ, correct=TRUE))
    # Missing values are not allowed and cannot be automatically corrected.
    wrong_econ <- okdata
    wrong_econ[3,"economic_zone"] <- "" # empty
    expect_warning(out <- check_format(wrong_econ))
    expect_warning(out <- check_format(wrong_econ, correct=TRUE))
    wrong_econ[3,"economic_zone"] <- as.character(NA) 
    expect_warning(out <- check_format(wrong_econ))
    expect_warning(out <- check_format(wrong_econ, correct=TRUE))
    wrong_econ[3,"economic_zone"] <- 56 # If numeric
    expect_warning(out <- check_format(wrong_econ))
    expect_warning(out <- check_format(wrong_econ, correct=TRUE))
})

test_that("rectangles", {
    # Remove extra punctuation - non-alphanumeric symbols
    wrong_rect <- okdata
    # with extra punctuation
    wrong_rect[3,"rectangle"] <- "39F0'" 
    expect_warning(out <- check_format(wrong_rect))
    expect_warning(corout <- check_format(wrong_rect, correct=TRUE), regexp=NA)
    expect_warning(out <- check_format(corout), regexp=NA)
    # Missing values are not allowed and cannot be automatically corrected.
    wrong_rect <- okdata
    wrong_rect[3,"rectangle"] <- ""
    expect_warning(out <- check_format(wrong_rect))
    expect_warning(out <- check_format(wrong_rect, correct=TRUE))
    wrong_rect[3,"rectangle"] <- as.character(NA)
    expect_warning(out <- check_format(wrong_rect))
    expect_warning(out <- check_format(wrong_rect, correct=TRUE))
    wrong_rect[3,"rectangle"] <- 666
    expect_warning(out <- check_format(wrong_rect))
    expect_warning(out <- check_format(wrong_rect, correct=TRUE))
})

test_that("trip identifier is unique", {
    # one trip, two days, different departure time, same identifier - no correction
    wrong_unique <- okdata
    wrong_unique[4,"deptime"] <- "0731"
    expect_warning(out <- check_format(wrong_unique))
    expect_warning(out <- check_format(wrong_unique, correct=TRUE))
    # Same trip identifier, but different vessel name
    wrong_unique <- okdata
    wrong_unique <- rbind(wrong_unique, okdata[1,])
    wrong_unique[nrow(wrong_unique),"eunr_id"] <- "other_boat"
    expect_warning(out <- check_format(wrong_unique))
    expect_warning(out <- check_format(wrong_unique, correct=TRUE))
})

test_that("duplicate rows", {
    # No duplicate rows allowed
    wrong_dup <- okdata
    # Add a duplicate row
    wrong_dup <- rbind(wrong_dup, wrong_dup[1,])
    expect_warning(out <- check_format(wrong_dup))
    expect_warning(outcor <- check_format(wrong_dup, correct=TRUE), regexp=NA)
    expect_warning(out <- check_format(outcor), regexp=NA)
    expect_equal(outcor, wrong_dup[-nrow(wrong_dup),])
})


R/data.R
# Author: Finlay Scott (EC JRC) <iago.mosqueira@jrc.ec.europa.eu>, Nuno Prista (SLU) and Thomas Reilly (Marine Scotland)
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

# Man pages for data sets

#' Fishing area
#'
#' A dataset of fishing areas given as DCF level 3 (DCF level 4 for the Baltic).
#'
#' @format A data frame with 150 rows and 2 variables:
#' \describe{
#'   \item{fishing_area}{code of the fishing area}
#'   \item{description}{description of the fishing area}
#' }
"fishing_areas"

#' Gear codes
#'
#' A dataset of gear codes taken from the Master Data Register.
#'
#' @format A data frame with 61 rows and 3 variables:
#' \describe{
#'   \item{gear_code}{code of the gear}
#'   \item{gear_type}{type of the gear}
#'   \item{description}{description of the gear}
#'   \item{passive}{is the gear passive (TRUE) or active (FALSE)}
#' }
"gear_codes"

#' ICES rectangles
#'
#' A dataset of ICES rectangles.
#'
#' @format A data frame with 7068 rows and 5 variables:
#' \describe{
#'   \item{ices_rectangle}{code of the rectangle}
#'   \item{south}{southern coordinate}
#'   \item{west}{westerly coordinate}
#'   \item{north}{northern coordinate}
#'   \item{east}{easterly coordinate}
#' }
"ices_rectangles"




R/calc_effort.R
# Author: Finlay Scott (EC JRC) <iago.mosqueira@jrc.ec.europa.eu>, Nuno Prista (SLU) and Thomas Reilly (Marine Scotland)
# Distributed under the terms of the European Union Public Licence (EUPL) V.1.1.

#' Calculate days at sea for a fishing trip.
#'
#' Calculate days at sea for a single fishing trip using data in the format
#' described in the package vignette \emph{checking_data}.
#'
#' The input is a single fishing trip. The format of the data should be
#' checked by \code{\link{check_format}} before calling this function 
#' (see the package vignette \emph{checking_data} for more details).
#' Days at sea is reported at the gear (type and mesh size), fishing area and
#' economic zone level. The total number of days at sea is the number of
#' commenced 24 hour periods.
#' The total number of days at sea of a trip is split equally over dates on
#' which fishing occurs. The effort for each fishing date is split equally over
#' the fishing activity on that date.  Active and passive gears are treated
#' equally.
#' See the vignette \emph{calculating_fishing_effort} for more details.
#' This function is called by \code{\link{calc_fishing_effort}}.
#'
#' @param trip Data.frame of the trip data
#' @return A \code{data.frame} with the days at sea by gear, fishing area
#' and economic zone.
#' @export
#' @seealso See \code{\link{calc_fishing_effort}}.
#' See the package vignette \emph{checking_data} for data preparation
#' and the vignette \emph{calculating_fishing_effort} for the calculation
#' details.
#' @examples
#' trip1 <- data.frame(
#'    eunr_id = "my_boat", loa = 2000, gt = 70, kw = 400,
#'    trip_id = "trip1",
#'    # 4 day trip
#'    depdate = "20140718", deptime = "0615", retdate = "20140721", rettime = "1615",
#'    # Only fish on 2 of those
#'    fishdate = c("20140719", "20140719", "20140719", "20140719", "20140720",
#'        "20140720", "20140720"), 
#'    gear = c("OTB","OTB","OTB","GN","OTB","GN","FPO"), gear_mesh_size = c(80,80,80,50,80,50,0),
#'    fishing_area = "27.4.B",
#'    economic_zone = "EU",
#'    rectangle = c("39F0","40F0","41F0","41F0","41F0","41F0","41F0"),
#'    stringsAsFactors = FALSE
#' )
#' das <- calc_days_at_sea_trip(trip1)
calc_days_at_sea_trip <- function(trip){
    # Check that this is just 1 trip
    if(length(unique(trip$trip_id)) != 1){
        stop("More than one trip found in data. This function only processes one trip at a time,")
    }
    # Calculate total days at sea
    # Number of commenced 24 hour periods - use lubridate interval class
    trip_start <- lubridate::ymd_hm(paste(trip$depdate[1],trip$deptime[1],sep=""))
    trip_end <- lubridate::ymd_hm(paste(trip$retdate[1],trip$rettime[1],sep=""))
    total_das <- ceiling(lubridate::int_length(lubridate::interval(trip_start, trip_end)) / (24*60*60))
    # Within each unique fish date, the days effort is split equally between fishing activity
    no_fishdates <- length(unique(trip$fishdate))
    # Add a column of the proportion of daily effort attributed to each activity
    fishdate <- NULL # Just to get rid of NOTE about visible binding when running check - seems annoying
    triptemp <- plyr::ddply(trip, "fishdate", transform, activity_prop = 1/length(fishdate))
    triptemp$days_at_sea <- triptemp$activity_prop * (total_das/no_fishdates)
    # Collapse date and rectangle
    days_at_sea <- NULL # Just to get rid of NOTE about visible binding when running check - seems annoying
    out <- plyr::ddply(triptemp, c("eunr_id", "loa", "gt", "kw", "trip_id", "depdate", "deptime", "retdate", "rettime", "gear", "gear_mesh_size", "fishing_area", "economic_zone"), plyr::summarise, days_at_sea = sum(days_at_sea))
    
    return(out)
}

#' Calculate fishing days for a fishing trip.
#'
#' Calculate fishing days for a single fishing trip using data in the format
#' described in the package vignette \emph{checking_data}.
#'
#' The input is a single fishing trip. The format of the data should be
#' checked by \code{\link{check_format}} before calling this function 
#' (see the package vignette \emph{checking_data} for more details).
#' Fishing days is reported at the gear (type and mesh size), fishing area, economic zone and rectangle level.
#' Passive and active gears are treated separately.
#' For active gears, each fishing date has 1 fishing day that is spread equally over the active gears.
#' For passive gears, each use of a passive gear is one fishing day, i.e. on fishing date can have several passive fishing days simultaneously.
#' See the vignette \emph{calculating_fishing_effort} for more details.
#' This function is called by \code{\link{calc_fishing_effort}}.
#'
#' @param trip Data.frame of the trip data
#' @return A data.frame with the fishing days by gear, fishing area, economic zone and rectangle.
#' @export
#' @seealso See \code{\link{calc_fishing_effort}}.
#' See the package vignette \emph{checking_data} for data preparation
#' and the vignette \emph{calculating_fishing_effort} for the calculation
#' details.
#' @examples
#' trip1 <- data.frame(
#'    eunr_id = "my_boat", loa = 2000, gt = 70, kw = 400,
#'    trip_id = "trip1",
#'    # 4 day trip
#'    depdate = "20140718", deptime = "0615", retdate = "20140721", rettime = "1615",
#'    # Only fish on 2 of those
#'    fishdate = c("20140719", "20140719", "20140719", "20140719", "20140720",
#'        "20140720", "20140720"), 
#'    gear = c("OTB","OTB","OTB","GN","OTB","GN","FPO"), gear_mesh_size = c(80,80,80,50,80,50,0),
#'    fishing_area = "27.4.B",
#'    economic_zone = "EU",
#'    rectangle = c("39F0","40F0","41F0","41F0","41F0","41F0","41F0"),
#'    stringsAsFactors = FALSE
#' )
#' fd <- calc_fishing_days_trip(trip1)
calc_fishing_days_trip <- function(trip){
    # Check that this is just 1 trip
    if(length(unique(trip$trip_id)) != 1){
        stop("More than one trip found in data. This function only processes one trip at a time,")
    }
    # Separate passive and active gears - gear_codes is a data set included in the package - lazy loading
    gear_codes <- fecR::gear_codes
    passive_gears <- gear_codes[gear_codes$passive,"gear_code"]
    active_gears <- gear_codes[!gear_codes$passive,"gear_code"]
    active <- trip[trip$gear %in% active_gears,]
    passive <- trip[trip$gear %in% passive_gears,]
    # Process active gears - date which has some fishing is given 1 fishing day. This is spread equally over all active fishing activities on that date.
    # Add a column of the fishing day attributed to each activity
    fishdate <- NULL # Just to get rid of NOTE about visible binding when running check - seems annoying
    activetemp <- data.frame()
    if (nrow(active)>0){
        activetemp <- plyr::ddply(active, "fishdate", transform, fishing_days = 1/length(fishdate), passive_gear = FALSE)
    }
    # Process passive gears - each entry is a passive gear used on a day so gets 1 fishing day
    passivetemp <- data.frame()
    if (nrow(passive)>0){
        passivetemp <- transform(passive, fishing_days=1, passive_gear = TRUE)
    }
    # Stick the results together
    triptemp <- rbind(activetemp, passivetemp)
    # Collapse date
    fishing_days <- NULL # Just to get rid of NOTE about visible binding when running check - seems annoying
    out <- plyr::ddply(triptemp, c("eunr_id", "loa", "gt", "kw", "trip_id", "depdate", "deptime", "retdate", "rettime", "gear", "gear_mesh_size", "rectangle", "fishing_area", "economic_zone"), plyr::summarise, fishing_days = sum(fishing_days))
    return(out)
}

#' Calculate days at sea and fishing days for a full trip data set.
#'
#' Calculates the days at sea and the fishing days for each trip in a data set.
#'
#' The input is a \code{data.frame} that contains details of fishing trips.
#' The format of the data should be checked by \code{\link{check_format}} before calling this function 
#' (see the package vignette \emph{checking_data} for more details on the data format).
#' See the documentation of \code{\link{calc_fishing_days_trip}} and \code{\link{calc_days_at_sea_trip}}
#' and the vignette \emph{calculating_fishing_effort} for more details
#' of how the different effort measures are calculated.
#' This function has the additional option of calling \code{\link{check_format}} before
#' the calculations are performed.
#'
#' @param dat data.frame with the details of all the fishing trips.
#' @param check_data Should the \code{\link{check_format}} function be called on the data first (default is TRUE).
#' It is not possible to run \code{\link{check_format}} with automatic corrections here. Do this yourself first.
#' @return A list with two data.frames: one with the fishing days by gear, fishing area, economic zone and rectangle,
#' the other with the days at sea by gear, fishing area and economic zone.
#' @export
#' @seealso See \code{\link{check_format}}.
#' See the package vignette \emph{checking_data} for data preparation
#' and the vignette \emph{calculating_fishing_effort} for the calculation
#' details.
#' @examples
#' trip1 <- data.frame(
#'     eunr_id = "my_boat", loa = 2000, gt = 70, kw = 400,
#'     trip_id = "trip1",
#'     # 4 day trip
#'     depdate = "20140718", deptime = "0615", retdate = "20140721", rettime = "1615",
#'     # Only fish on 2 of those
#'     fishdate = c("20140719", "20140719", "20140719", "20140719", "20140720",
#'        "20140720", "20140720"), 
#'     gear = c("OTB","OTB","OTB","GN","OTB","GN","FPO"), gear_mesh_size = c(80,80,80,50,80,50,0),
#'     fishing_area = "27.4.B",
#'     economic_zone = "EU",
#'     rectangle = c("39F0","40F0","41F0","41F0","41F0","41F0","41F0"),
#'     stringsAsFactors = FALSE
#' )
#' trip2 <- data.frame(
#'     eunr_id = "my_boat", loa = 2000, gt = 70, kw = 400,
#'     trip_id = "trip2",
#'     # 2 day trip
#'     depdate = "20140718", deptime = "0615", retdate = "20140719", rettime = "0600",
#'     # Only fish on 2 of those
#'     fishdate = c("20140718", "20140719", "20140719", "20140719"), 
#'     gear = c("OTB","OTB","GN","FPO"), gear_mesh_size = c(80,80,50,0),
#'     fishing_area = "27.4.B",
#'     economic_zone = "EU",
#'     rectangle = c("39F0","39F0"),
#'     stringsAsFactors = FALSE
#' )
#' dat <- rbind(trip1, trip2)
#' effort <- calc_fishing_effort(dat)
calc_fishing_effort <- function(dat, check_data = TRUE){
    # Check data first of all
    if(check_data){
        print("Running check on data before calculating effort")
        dat <- tryCatch(check_format(dat),error=function(e) e, warning=function(w) w)
        if(methods::is(dat, "warning")){
            stop("Checking the input data generated a warning. Stopping here")
        }
        print("Data looks OK. Continuing")
    }
    # Get days at sea
    das <- plyr::ddply(dat, "trip_id", calc_days_at_sea_trip)
    fd <- plyr::ddply(dat, "trip_id", calc_fishing_days_trip)
    return(list(days_at_sea = das, fishing_days = fd))
}




